scene/2d/tile_map.h

```diff
@@ -89,75 +89,134 @@ class TerrainConstraint {
-			// We sort the cells by their local coords, as it is needed by rendering.
-			if (p_a.y == p_b.y) {
-				return p_a.x > p_b.x;
-			} else {
-				return p_a.y < p_b.y;
-			}
-		}
-	};
-
-	// Dirty list element.
-	SelfList<TileMapQuadrant> dirty_list_element;
+struct DebugQuadrant;
+struct RenderingQuadrant;
 
-	// Quadrant coords.
+struct CellData {
 	Vector2i coords;
-
-	// TileMapCells.
-	RBSet<Vector2i> cells;
-	// We need those two maps to sort by local position for rendering.
-	// This is kind of workaround, it would be better to sort the cells directly in the "cells" set instead.
-	RBMap<Vector2i, Vector2> map_to_local;
-	RBMap<Vector2, Vector2i, CoordsWorldComparator> local_to_map;
+	TileMapCell cell;
 
 	// Debug.
-	RID debug_canvas_item;
+	SelfList<CellData> debug_quadrant_list_element;
 
 	// Rendering.
-	List<RID> canvas_items;
-	HashMap<Vector2i, RID> occluders;
+	Ref<RenderingQuadrant> rendering_quadrant;
+	SelfList<CellData> rendering_quadrant_list_element;
+	List<RID> occluders;
 
 	// Physics.
-	List<RID> bodies;
+	LocalVector<RID> bodies;
 
 	// Navigation.
-	HashMap<Vector2i, Vector<RID>> navigation_regions;
+	LocalVector<RID> navigation_regions;
 
 	// Scenes.
-	HashMap<Vector2i, String> scenes;
+	String scene;
 
 	// Runtime TileData cache.
-	HashMap<Vector2i, TileData *> runtime_tile_data_cache;
-
-	void operator=(const TileMapQuadrant &q) {
-		coords = q.coords;
-		debug_canvas_item = q.debug_canvas_item;
-		canvas_items = q.canvas_items;
-		occluders = q.occluders;
-		bodies = q.bodies;
-		navigation_regions = q.navigation_regions;
+	TileData *runtime_tile_data_cache = nullptr;
+
+	// List elements.
+	SelfList<CellData> dirty_list_element;
+
+	// For those, copy everything but SelfList elements.
+	void operator=(const CellData &p_other) {
+		coords = p_other.coords;
+		cell = p_other.cell;
+		occluders = p_other.occluders;
+		bodies = p_other.bodies;
+		navigation_regions = p_other.navigation_regions;
+		scene = p_other.scene;
+		runtime_tile_data_cache = p_other.runtime_tile_data_cache;
 	}
-
-	TileMapQuadrant(const TileMapQuadrant &q) :
+	CellData(const CellData &p_other) :
+			debug_quadrant_list_element(this),
+			rendering_quadrant_list_element(this),
 			dirty_list_element(this) {
-		coords = q.coords;
-		debug_canvas_item = q.debug_canvas_item;
-		canvas_items = q.canvas_items;
-		occluders = q.occluders;
-		bodies = q.bodies;
-		navigation_regions = q.navigation_regions;
+		coords = p_other.coords;
+		cell = p_other.cell;
+		occluders = p_other.occluders;
+		bodies = p_other.bodies;
+		navigation_regions = p_other.navigation_regions;
+		scene = p_other.scene;
+		runtime_tile_data_cache = p_other.runtime_tile_data_cache;
 	}
 
-	TileMapQuadrant() :
+	CellData() :
+			debug_quadrant_list_element(this),
+			rendering_quadrant_list_element(this),
 			dirty_list_element(this) {
 	}
 };
 
+struct DebugQuadrant : public RefCounted {
+	GDCLASS(DebugQuadrant, RefCounted);
+
+public:
```

Comment:
This is a `struct`, so it's already public.
Not sure why didn't you use `class` though.